ui/pages/confirmations/components/confirm/info/typed-sign/typed-sign-v4-simulation/decoded-simulation/decoded-simulation.tsx

```diff
@@ -104,19 +106,39 @@ const DecodedSimulation: React.FC<object> = () => {
   const chainId = currentConfirmation.chainId as Hex;
   const { decodingLoading, decodingData } = currentConfirmation;
 
-  const stateChangeFragment = useMemo(
-    () =>
-      (decodingData?.stateChanges ?? []).map(
-        (change: DecodingDataStateChange) => (
+  const stateChangeFragment = useMemo(() => {
+    const stateChangesGrouped: Record<string, DecodingDataStateChange[]> = (
+      decodingData?.stateChanges ?? []
+    ).reduce(
+      (
+        result: Record<string, DecodingDataStateChange[]>,
```

Comment:
Minor, you can pass the target type as a template on `reduce` and you won't need to cast the output and `result` variable.